kitsune/middleware.py

```diff
@@ -0,0 +1,199 @@
+# -*- coding: utf-8 -*-
+#
+# This tries to fix bug 1357563 in a best effort way. It may find a proper page
+# to redirect users to but if it fails it will raise a 404.
+#
+# Flow:
+#
+# 1. Check if the response is 404 and it's from a Lithium URL (/t5/*)
+# 2. Checks cache, redirects if successful
+# 3. Checks Lithium localized product_pages from LithiumRedirectionMiddleware.product_pages, redirects if successful
+# 4. Fetches the page from Lithium and tries to extract title of the page.
+#    Use og:title instead of HTML Title because Lithium strips the later down to 50 chars or something.
+#    If it fails it returns 404
+# 5a. Try to find a Document with the same title, redirect if successful
+# 5b. Try to find a Topic with the same title under a product. Product is also extracted by title. For non-English pages we need to do a reverse translation lookup from TOPICS
+# 5c. Try to find a Product. For non-English pages we need to do a reverse translation lookup from PRODUCTS
+# 5d. If everything fails redirect to the search page with title as search string.
+#
+# In all cases we store a successful redirect in the cache.
+# In all cases we append ?cache=no in the redirect to bust the cache
+# All redirects are 302s
+#
+#
+
+import hashlib
+import re
+import urllib
+
+from django.core.cache import cache
+from django.conf import settings
+from django.http import HttpResponseRedirect
+from django.utils.translation import pgettext
+from django.utils import translation
+
+import requests
+from raven.contrib.django.raven_compat.models import client as sentry_client
+
+from kitsune.products.models import Product, Topic
+from kitsune.sumo.views import handle404
+from kitsune.wiki.models import Document
+
+
+# build a list of topics translated for reverse lookup
+TOPICS = {}
+for topic in set(Topic.objects.values_list('title', flat=True)):
+    for lang in settings.SUMO_LANGUAGES:
+        with translation.override(lang):
+            trans_text = pgettext('DB: products.Topic.title', topic)
+        TOPICS[trans_text] = topic
+
+PRODUCTS = {}
+for topic in set(Product.objects.values_list('title', flat=True)):
+    for lang in settings.SUMO_LANGUAGES:
+        with translation.override(lang):
+            trans_text = pgettext('DB: products.Product.title', topic)
+        PRODUCTS[trans_text] = topic
+
+
+class LithiumRedirectionMiddleware():
+    """
+    Redirect requests for user profiles from /<username> to
+    /u/<username> to avoid breaking profile urls with the new url
+    schema.
```

Comment:
Is this the right docstring for this class? It doesn't seem to match anything else around here.